crates/burn-autodiff/src/runtime/memory_management.rs

```diff
@@ -115,81 +115,109 @@ impl GraphMemoryManagement {
-                    .expect("All nodes should have received a status at this point")
-                    .clone();
-
-                match node_status {
-                    NodeMemoryStatus::Useful => {
-                        // Nothing to do, was already tagged through some other path
-                    }
-                    NodeMemoryStatus::Unavailable => {
-                        // Even if this node is unavailable, it is still possible that an ancestor is useful if referenced
-                        for parent in parents {
-                            self.useful_propagation(parent, Mode::Explore);
+                Mode::Explore => {
+                    let node_status = self
+                        .statuses
+                        .get(&node_id)
+                        .expect("All nodes should have received a status at this point")
+                        .clone();
+
+                    match node_status {
+                        NodeMemoryStatus::Useful => {
+                            // Nothing to do, was already tagged through some other path
+                            continue;
                         }
-                    }
-                    NodeMemoryStatus::Unknown => {
-                        // If this node is referenced and not unavailable,
-                        // then it is useful and we must retain all ancestors
-
-                        let mut mode = Mode::Explore;
-                        if self.is_referenced(node_id) {
-                            self.statuses.insert(node_id, NodeMemoryStatus::Useful);
-                            mode = Mode::TagAsUseful;
+                        NodeMemoryStatus::Unavailable => {
+                            // Even if this node is unavailable, it is still possible that an ancestor is useful if referenced
+                            Mode::Explore
                         }
-
-                        for parent in parents {
-                            self.useful_propagation(parent, mode.clone());
+                        NodeMemoryStatus::Unknown => {
+                            // If this node is referenced and not unavailable,
+                            // then it is useful and we must retain all ancestors
+                            if self.is_referenced(node_id) {
+                                self.statuses.insert(node_id, NodeMemoryStatus::Useful);
+                                Mode::TagAsUseful
+                            } else {
+                                Mode::Explore
+                            }
                         }
                     }
                 }
+            };
+
+            for parent in self
+                .nodes
+                .get(&node_id)
+                .cloned()
+                .unwrap_or(vec![])
+                .into_iter()
+            {
+                if !visited.contains(&parent) {
+                    to_visit.push((parent, next_mode.clone()));
+                }
```

Comment:
As I understand it, we can still visit multiple times the same parent with different modes, but once we visited the parent once, we can't register new modes. Wouldn't it make sense to register the modes as well with priority (If `TagAsUseful` > `Explore`)?

The `visited` could be an `HashMap` containing the mode used. You can't register a parent with `Explore` when the parent was already visited, but you can with the mode `TagAsUseful` if the parent was visited with the mode `Explore`?

Reply:
Unlike the NodeMemoryStatus for which there is one status per node (at first Unknown, then maybe Unavailable after the first propagation, then maybe Useful after this propagation), there is not one Mode per node. 

The Mode made more sense in the previous form of the algorithm: we started in exploration until we found a node to tag as useful, then for this node to be usable we had to tag all of its parents as useful as well, so the algorithm switched to TagAsUseful for all of this branch. But with the `visited` approach we have to remember to go back to this mode when we get to a parent of a useful node. So the mode becomes tied to nodes in the vec and it's not very elegant as it was supposed to be a status of the algorithm, not the node. I think I can remove the concept of Mode altogether, it will be clearer.